ROFO-59 로그인 및 인증 필터 구현

Unagi-zoso opened the pull request:
### 개요
회원가입 기능이 구현되어 로그인 및 인증 필터를 구현해야 합니다.

### 변경사항
- JJWT 라이브러리 종속성 추가
- JwtProperties, JwtUtil 작성
- JwtFilter 작성
- LoginUserResolver 작성
- 테스트용 SecurityConfig 분리
- 로그인, 토큰갱신, 로그아웃 API 작성

### 테스트
- 테스트 코드 추가여부 O

### 관련 지라 및 위키 링크
 - [JIRA](https://weit-2nd.atlassian.net/browse/ROFO-59) 


### 리뷰어에게 하고 싶은 말
모든 피드백 감사히 잘 받겠습니다!
로그인, 토큰갱신, 로그아웃 기능은 나름 심혈을 기울여 만들긴했지만
인증필터를 구현하게 되면서 테스트 환경에서의 인증과정도 생각해볼 수 밖에 없었습니다.
이번에 구현한 인증필터(JwtFilter)를 그대로 테스트환경에서 적용시키면 @WebMvcTest 를 할 때 필터가 의존하고 있는
빈들을 Mock 객체로 두고 스터빙을 하고 등등 부가적인 코드가 너무 많이 필요했습니다.
좋은 방법이 없을까 생각하다. 프로덕트, 개발 환경과 테스트 환경에서의 SecurityConfig 를 달리해
테스트 환경에선 다른 객체를 의존하지 않는 기존 MockPassFilter 를 사용하게 하는 것이 였습니다.
그러면 JwtFilter 코드에 대한 신뢰가 떨어지지 않을까 걱정이 들어 나름 JwtFilter 를 테스트하기 위한 코드 또한
조금 작성해봤습니다. 이 부분도 개선할 방법이 있다면 피드백 부탁드리겠습니다. 
감사합니다!

Unagi-zoso committed:
Merge branch 'develop' into feature/ROFO-59

Unagi-zoso committed:
Merge branch 'develop' into feature/ROFO-59

Unagi-zoso committed:
ROFO-59 JJWT 라이브러리 종속성 추가

Unagi-zoso committed:
ROFO-59 JwtProperties, JwtUtil, HMacKeyGenerator 클래스 작성

Unagi-zoso committed:
ROFO-59 JwtFilter 및 관련 클래스 작성

Unagi-zoso committed:
ROFO-59 UserRepository getBySocialId 메소드 추가

Unagi-zoso committed:
ROFO-59 테스트용 SecurityConfig 분리

기존 SecurityConfig 에 JwtFilter 가 적용되는데 이때 JwtUtil, CustomUserDetailService 같은 빈을 의존한다.
이를 위해 매번 MockBean 으로 설정하고 stubbing 을 해야한다.
이 부분이 번거로워 테스트 환경에선 이전에 사용하던 MockPassFilter 를 통해 검증을 하게 한다.

Unagi-zoso committed:
ROFO-59 LoginUserResolver 작성

Unagi-zoso committed:
ROFO-59 ServiceTokensResponse 및 AuthFixture 추가

Unagi-zoso committed:
ROFO-59 login, reissueTokens, logout 메소드 및 테스트 작성

Unagi-zoso committed:
ROFO-59 회원가입, 토큰갱신, 로그아웃 API 작성

Unagi-zoso committed:
ROFO-59 필터단 인증 실패 시 예외메시지 변경

너무 삭막하게 느껴져 변경했습니다.

Unagi-zoso committed:
ROFO-59 스웨거 시큐리티 userId 부분 제거

Unagi-zoso committed:
ROFO-59 로그인 API 예외 메시지 수정

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthController.kt:
```diff
@@ -39,12 +44,38 @@ class AuthController(
+        requireNotNull(refreshToken) { "RefreshToken 이 존재하지 않습니다." }
+        if (refreshToken.isNotBearerToken()) {
+            throw IllegalArgumentException("RefreshToken 의 형식이 올바르지 않습니다.")
+        }
```
여긴
```require(refreshToken.isNotBearerToken()) {"RefreshToken 의 형식이 올바르지 않습니다."}```
으로 쓰면 좀더 코틀린스러운 코드가 가능할것 같아요~!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/spec/AuthControllerSpec.kt:
```diff
@@ -178,4 +180,199 @@ interface AuthControllerSpec {
+                description = "로그인 성공",
+            ),
+            ApiResponse(
+                responseCode = "400",
```
이것도 401로 내려가는게 맞지 않을까 싶은데 어떻게 생각하시나요??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/security/CustomUserDetailService.kt:
```diff
@@ -0,0 +1,49 @@
+class CustomUserDetailService(
+    private val userRepository: UserRepository,
+) :
+    UserDetailsService {
```
뭔가 포멧팅이 이상한데 ktlint가 제대로 돌아가고 있는지 확인한번 부탁드립니다~!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/security/CustomUserDetailService.kt:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+class SecurityUser(val user: User) : UserDetails {
```
user엔티티가 UserDetails를 상속받아서 사용하는것만 봤는데
요런식으로 user를 랩핑하는 별도의 클래스가 있는것도 나쁘지 않네요 ㅋㅋㅋㅋ

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/security/config/SecurityConfig.kt:
```diff
@@ -20,25 +26,24 @@ class SecurityConfig(
     arrayOf(
         "/health",
         "/ready",
-        "/api/v1/test/success",
-        "/api/v1/test/error",
-        "/api/v1/test/rest",
```
아 ㅋㅋㅋㅋ 제가 이걸 빼먹었군요

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/security/jwt/JwtProperties.kt:
```diff
@@ -0,0 +1,11 @@
+
+@ConfigurationProperties(prefix = "jwt")
+class JwtProperties(
+    val accessKey: String = generateHmac256Key(),
+    val refreshKey: String = generateHmac256Key(),
+    val accessTokenExpirationTime: Long = 5 * 60 * 1000,
+    val refreshTokenExpirationTime: Long = 30 * 60 * 1000,
```
오잉 요거 잘 이해가 안되는데 미리정의한 프로퍼티를 쓰는게 아니라 매 실행마다 동적으로 생성하는건가요??

Unagi-zoso replied:
제가 ParameterStore 를 잘 안써봐서 잘못 사용하는거일 수 있습니다.
그래도 제가 의도한 바를 말씀드리자면 ParameterStore 에 의한 **민감한 application.yml 들의 주입**이
**sandbox 배포환경에서만** 일어난다면 이를 사용하는 로컬 테스트 환경에서 이 값들을 찾지 못해
인스턴스 생성에 실패할 것이라 생각했습니다. 
따라서 위와 같이 기본값들을 정의해 **로컬환경**에서는 
**임의의 기본값**들과 **실제 배포환경**에서는 **ParameterStore 에서 주입 받은 값**들로 돌아가게 의도하였습니다.
`application.yml` 에 `jwt` 를 import 하게 추가했는데 이것도 배포환경에서 주입될 applicatin-jwt.yml 을 위한 것입니다.

개발환경에서 테스트해본 결과도 제가 의도한 대로 적용되는 것 같습니다.

혹시 제가 잘못 사용하고 있다면 조금 알려주시면 감사하겠습니다!

Unagi-zoso replied:
application-jwt.yml 을 test profile 에만 정적으로 데이터를 가져와 사용할 수도 있지만
매번 새로운 key 값을 동적으로 생성하는 것도 괜찮아보여서 적용해봤습니다

konempty replied:
> 매번 새로운 key 값을 동적으로 생성하는 것도 괜찮아보여서 적용해봤습니다

요렇게 쓰면 서버 재기동할때마다 JWT유효성 검증을 하는 키가 달라져서 까다로워질것 같아요..!

konempty replied:
>sandbox 배포환경에서만 일어난다면 이를 사용하는 로컬 테스트 환경에서 이 값들을 찾지 못해
인스턴스 생성에 실패할 것이라 생각했습니다.

따로 명시하지 않는다면 로컬에서도 Sandbox 페이즈로 돌아가도록 되어있어서 파라메터 스토어에 추가되는 내용들이
동일하게 적용됩니다~!
자세한 내용은 Spring Cloud Config를 공부해보시면 이해하실수 있을것 같아요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/security/jwt/JwtUtil.kt:
```diff
@@ -0,0 +1,141 @@
+    ): String =
+        generateJwtToken(
+            mapOf(
+                "socialId" to socialId,
```
여기에 유저번호도 내려줘서 자기의 번호가 몇번인지 알도록하는건 어떨까요??

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthController.kt:
```diff
@@ -39,12 +44,38 @@ class AuthController(
         profileImage: MultipartFile?,
     ) {
-        requireNotNull(socialAccessToken) { "socialAccessToken 가 존재하지 않습니다." }
+        requireNotNull(socialAccessToken) { "socialAccessToken 이 존재하지 않습니다." }
```
socialAccessToken 이 존재하지 않습니다.

SocialAccessToken 이 존재하지 않습니다.

중복되는 매직 리터럴인데
ErrorCode 내부에서 선언하는건 어떨까요?

Unagi-zoso replied:
좋은 의견 감사합니다! 
제가 구현하는 기능에선 저 예외 메시지의 재사용 빈도가 그리 자주 있진 않았습니다.
회원가입, 로그인 두 상황에서 주로 사용된다고 판단하고 있습니다.
이후에도 이 부분을 변경할 일이 크게 없을 것 같습니다.
그래서 미리 ErrorCode 에 선언하기 보단 현 상태를 유지하고 재사용되는 상황이 찾아온다면
그때 꼭 반영해보겠습니다!

jifrozen0110 replied:
넵 알겠습니다!
그럼 위 아래문구가
socialAccessToken
SocialAccessToken
한개는 대문자이고 한개는 소문자입니다 통일하는것이 어떨까요?

Unagi-zoso replied:
이럴 수가 정말 눈썰미가 좋으시네요.. 
요건 상태번호도 바꿀겸 커스텀 예외 객체로 다뤄야겠어요. 
감사합니다!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/fixture/AuthFixture.kt:
```diff
@@ -3,4 +3,8 @@ package kr.weit.roadyfoody.auth.fixture
+
+const val TEST_ROTATE_ID = "123456789"
+
+const val TEST_INVALID_ACCESS_TOKEN = "i.nv.alid.tok.en"
```
ㅋㅋㅋㅋㅋ 재미있네요

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthController.kt:
```diff
@@ -39,12 +44,38 @@ class AuthController(
+        return authCommandService.reissueTokens(refreshToken.removeBearer())
+    }
+
+    @PostMapping("/sign-out")
+    override fun signOut(
+        @LoginUser user: User,
+    ) {
+        authCommandService.logout(user)
+    }
```
반환할 데이터가 없어보이는데
204 (No Contents)를 내리는게 어떨까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/support/config/TestSecurityConfig.kt:
```diff
@@ -0,0 +1,63 @@
+            }.build()
+}
+
+class MockPassFilter : OncePerRequestFilter() {
+    override fun doFilterInternal(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        filterChain: FilterChain,
+    ) {
+        if (request.headerNames.toList().contains("userid")) {
+            val socialId = TEST_SOCIAL_ID
+            val securityUser = SecurityUser(User.of(socialId = "$TEST_SOCIAL_LOGIN_TYPE $socialId", nickname = TEST_USER_NICKNAME))
+            val authentication = UsernamePasswordAuthenticationToken(securityUser, null, securityUser.authorities)
+            SecurityContextHolder.getContext().authentication = authentication
+        }
+        filterChain.doFilter(request, response)
+    }
+}
```
테스트 할때는 JWT토큰을 통한 로그인처리가 아닌
userId 헤더를 통한 로그인 처리를 한다는 의도인거죠??
실제 환경과 다르니까 테스트로 검출되지 못한 에러가 발생할까봐 걱정이 되긴 하는데
요 부분이 문제 되면 그때 테스트에서도 JWT를 도입할지 고민해보도록 하시죠 ㅋㅋㅋ

konempty commented on src/main/resources/application.yml:
```diff
@@ -8,6 +8,7 @@ spring:
       - flyway
       - docs
       - api
+      - jwt
```
application-jwt.yml이 추가되지 않은것 같은데 요 라인을 추가하신 이유가 있을까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -54,6 +61,7 @@ class AuthIntegrationServiceTest(
                     }
                 }
             clearAllMocks()
+            redisTemplate.opsForValue().operations.delete(getObjectStorageCacheKey(TEST_USER_SOCIAL_ID))
```
요기는 
```kotlin
redisTemplate.delete(getObjectStorageCacheKey(TEST_USER_SOCIAL_ID))
```
요렇게 줄일수 있지 않을까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthControllerTest.kt:
```diff
@@ -126,16 +131,23 @@ class AuthControllerTest(
+                    verify(exactly = 0) {
+                        authCommandService.register(any<String>(), any<SignUpRequest>(), any<MultipartFile>())
+                    }
+                    verify {
+                        mockFile.size
+                        mockFile.name
+                        mockFile.inputStream
+                    }
```
요건 뭘 검증하기위한 코드일까요??

konempty reviewed (approved):
고생하셨습니다 👍 👍 
역시 웰노운 기능이라 그런지 코드가 깔끔하고 좋네요 ㅋㅋㅋ
코멘트 남겼으니까 반영 부탁드립니다~!

jifrozen0110 reviewed (approved):
고생하셨습니다! 👍 코드양이 대박이네요!

lin-chae commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandServiceTest.kt:
```diff
@@ -135,4 +148,112 @@ class AuthCommandServiceTest : BehaviorSpec({
+            then("ServiceTokensResponse 가 반환된다.") {
+                val actual = authCommandService.reissueTokens(TEST_BEARER_REFRESH_TOKEN)
+                actual.accessToken shouldBe TEST_ACCESS_TOKEN
+                actual.refreshToken shouldBe TEST_REFRESH_TOKEN
```
중요한 작업인만큼 테스트도 꼼꼼하게 짜주셨네요!

lin-chae reviewed (approved):
회원 기능이 잘 알려진 기능이지만 온갖 설정에 정말 힘든 작업이셨을텐데 정말 고생하셨습니다!

Unagi-zoso committed:
ROFO-59 application-jwt.yml 작성 및  JwtProperties 기본값 제거

Unagi-zoso committed:
ROFO-59 getBySocialid 메소드 제거

UserNotFoundException 을 요구하는 사용처가 없고
예외의 타입을 요구에 따라 바꾸기엔 요구하는 예외가 repository 단에 노출되지 않았음해서 제거하고 findBySocialId 메소드를 사용합니다.

Unagi-zoso committed:
ROFO-59 Jwt 토큰 Claim 중 socialId 제거 및 userId 추가

정수형의 userId 가 조회성능이 더 뛰어날 것으로 판단해 socialId 를 제거하고 userId 로 대체하였습니다.

Unagi-zoso committed:
ROFO-59 MissingSocialAccessTokenException 추가

Unagi-zoso committed:
ROFO-59 JwtFilter socialId 제거 후 userId 응용 로직 추가

Unagi-zoso committed:
ROFO-59 서비스단 변경사항 적용 및 피드백 반영

Unagi-zoso committed:
ROFO-59 컨트롤러단 피드백 반영

Unagi-zoso committed:
ROFO-59 테스트 코드 주석 수정

Unagi-zoso committed:
ROFO-59 테스트 코드 주석 수정
